documentation/solution-parser.md

```diff
@@ -0,0 +1,91 @@
+# Solution Parser
+
+## Expander
+
+`Expander` class is used to expand value from an expression, which can
+be function invocation. These expression (or embedded properties)
+have form - "$(property)".
+
+`Expander` handles intrinsic functions, which is a list of built-in functions
+(bunch of ifs).
+
+MSBuild defines initial set of properties, like `MSBuildFrameworkToolsPath32`,
+`MSBuildFrameworkToolsPath`, etc.
+
+Tools configuration can come from configuration file, registry and local.
+
+## Building
+
+Function `BeginBuild` prepares `BuildManager` to receive build requests, which:
+- Attaches debugger;
+- Checks that current build manager's state is idle;
+- Initializes logging service;
+- Initializes caches;
+- Registers packet handlers.
+
+After this setup, `MSBuild` creates build request data from project
+files or project instances. Data for build requests is stored in `BuildRequestData` class.
+
+After build data for request has been prepared, `MSBuild` executes build (notice that
+it can also execute restore and graphbuild depending on configuration). It executes
+build in `ExecuteBuild` function. This function pends build request, which creates
+an instance of `BuildSubmission` that represent build submission.
```

Comment:
```suggestion
After build data for a request have been prepared, `MSBuild` executes the build. It may execute restore and/or a graph build instead depending on the configuration. It executes the build in the `ExecuteBuild` function, which pends the build request, creating an instance of `BuildSubmission` that represents the build submission.
```